### ISSUE TITLE
Connections should be shared between OkHttpClients when using shared ConnectionPool

### ISSUE THREAD

**Issue by 15characterlimi** [square/okhttp#2757]
What kind of issue is this?

 - [X] Bug report.

## tl;dr
CertificateChainCleaner / CertificatePinner don't override equals() even though Platform.buildCertificateChainCleaner() constructs new (not interned) instances. This breaks connection pooling across separately constructed OkHttpClient instances with shared ConnectionPool.

## Failing tests
   - [Gist with a single failing test (adopted from a part of URLConnectionTest)](https://github.com/15characterlimi/okhttp/commit/119873249d7414964a0ea24fac79081ad7c7b7d9)
   - [Draft commit that adds two failing tests](https://github.com/15characterlimi/okhttp/commit/119873249d7414964a0ea24fac79081ad7c7b7d9)

## Detail

 - A few places (namely OkHttpClient(Builder), and both overloads of 
   OkHttpClient.Builder.sslSocketFactory()s), call
   CertificateChainCleaner.get(X509TrustManager) to initialize a
   OkHttpClient's CertificateChainCleaner, and
 - CertificateChainCleaner.get(X509TrustManager) always returns a new
   CertificateChainCleaner instance, and
 - CertificateChainCleaner.equals() is based on Object identity, and
 - CertificatePinner.withCertificateChainCleaner() returns a new CertificatePinner
   instance if the CertificateChainCleaner instances are different, and
 - Address.equals() depends on equal(this.certificatePinner, that.certificatePinner), and
 - ConnectionPool.get() only reuses a Connection if address.equals(connection.route().address)

The issue should be easy to work around once one finds the root cause, but in case I feel inspired to pull together a fix for you, what's your view on how this should be fixed?

I suggest that CertificateChainCleaner and CertificatePinner should implement value equality (i.e. override Object.equals()). That CertificatePinner.withCertificateChainCleaner() uses != to compare object instances (namely CertificateChainCleaner instances) seems like a major antipattern to me.

Alternatively, CertificateChainCleaner instances could be interned.

Both of those approaches would rely on X509TrustManager.equals() to work in order to figure out when a new CertificateChainCleaner instance is required; but from what see, X509TrustManager seems to be a singleton in practice so its equals() should not be a problem. 

I know the docs suggest using a OkHttpClient singleton instance as the base for all connections as well as additional OkHttpClient.Builders, however it seems like poor behavior from the library to rely on this in order for ConnectionPool instances to be shared.

**Comment by swankjesse** [comment]
Good find. Can fix.

**Comment by 15characterlimi** [comment]
I've sent a pull request to fix this, based on overriding equals() and hashCode() to implement value equality semantics on the certificate related classes. The complication is that this is easy to break (although the tests should now hopefully catch it). I'll wait to hear from you whether you like that approach or want to implement a different one.
